added support for tmpfs-mode in compose file

ethan-haynes opened the pull request:
Provides a fix for https://github.com/docker/cli/issues/698

**- What I did**
  Added the ability to use a tmpfs volume type in compose with the optional parameter of size. 
**- How I did it**
  Changed compose schemas 3.2-3.5 to allow for the type. Tweaked volumes parser to read it. 
  Removed it from unapproved params. 
**- How to verify it**

Test it with compose:

```yaml
version: "3.6"
services:
  tmpfs:
    image: nginx:latest
    volumes:
      - type: tmpfs
        target: /app
        tmpfs:
          size: 10000
```
```
$ docker stack deploy -c swarm-compose.yml nginx
```

thaJeztah commented:
design looks good to me as well; moving to code review

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+	result mount.Mount,
+) (mount.Mount, error) {
+	if volume.Volume != nil {
+		return result, errors.New("volume options are incompatible with type bind")
```
On error-situations, we should not return a `result` with values; the result should be "empty" Mount, e.g.;

```go
return mount.Mount{}, errors.New("volume options are incompatible with type bind")
```

but also see my other comment on moving the `result` mount outside of these functions; it's likely better to construct the result at the end of each `handleXX` function

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+
+	switch volume.Type {
+	case "volume":
+		mount, err := handleVolumeToMount(volume, stackVolumes, namespace, result)
```
no need for the intermediate variables here; this can just be;

```go
return handleVolumeToMount(volume, stackVolumes, namespace, result)
```

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+		mount, err := handleVolumeToMount(volume, stackVolumes, namespace, result)
+		return mount, err
+	case "bind":
+		mount, err := handleBindToMount(volume, result)
```
same here;

```go
return handleBindToMount(volume, result)
```

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+		mount, err := handleBindToMount(volume, result)
+		return mount, err
+	case "tmpfs":
+		mount, err := handleTmpfsToMount(volume, result)
```
and here;

```go
return handleTmpfsToMount(volume, result)
```

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+		mount, err := handleTmpfsToMount(volume, result)
+		return mount, err
+	}
+	return result, nil
```
This doesn't look right: unknown `volume.Type` should produce an error, and return an empty result

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+	stackVolumes volumes,
+	namespace Namespace,
+) (mount.Mount, error) {
+	result := mount.Mount{
```
I'm not a big fan of moving this outside of the `handleXX` functions; I assume you kept it out to reduce code duplication, but  now the conversion (and validation) is "split" between the `handleXX` functions and this function.

Given that not all mount-types use the same options in this type (e.g., `type=bind` uses `Source`, but `type=tmpfs` doesn't), I'd move this code to each `handleXX` function. If duplication of code is a concern (but I think it's minimal), you _could_ have a helper function for the shared code

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+			Propagation: mount.Propagation(volume.Bind.Propagation),
+		}
+	}
+	// Binds volumes
```
This comment doesn't add much, so I'd remove it

thaJeztah commented on cli/compose/convert/volume.go:
```diff
@@ -88,3 +67,77 @@ func convertVolumeToMount(
+			SizeBytes: volume.Tmpfs.Size,
+		}
+	}
+	// Tmpfs volumes
```
This comment doesn't add much, so I'd remove it; if the function needs documentation, add a GoDoc for the function itself

thaJeztah commented on cli/compose/schema/data/config_schema_v3.5.json:
```diff
@@ -300,6 +300,12 @@
+                  "tmpfs": {
+                    "type": "object",
+                    "properties": {
+                      "size": {"type": "number"}
```
should this be `"type": "integer"` ?

If only positive integers are allowed, likely:

```json
{
  "type": "integer",
  "exclusiveMinimum": 0
}
```

(not sure if `exclusiveMinimum` or `minimum`; likely depends on how empty values are handled)

ethan-haynes replied:
In the Docker [documentation](https://docs.docker.com/engine/admin/volumes/tmpfs/#specify-tmpfs-options) if tmpfs-size isn't provided it will be unlimited by default:

Option | Description
-- | --
tmpfs-size | Size of the tmpfs mount in bytes. Unlimited by default.

So if the user wants to specify a ceiling of how many bytes to allocate, then they should provide a positive integer above 0. The docker cli, as it behaves now, will not allow negative numbers:
```
invalid argument "type=tmpfs,destination=/app,tmpfs-size=-1" for --mount: invalid value for tmpfs-size: -1
```
It will allow zero, but when you inspect the container it just defaults TmpfsOptions to an empty object, which (since tmpfs-size won't actually be set) will allocate "unlimited" resources: 
```JSON
"Mounts": [
                        {
                            "Type": "tmpfs",
                            "Target": "/app",
                            "TmpfsOptions": {}
                        }
                    ]
```

So I think we should, as you said, use an exclusiveMinimum of zero. That way the user at least attaches the TmpfsOptions with a value for tmpfs-size.

thaJeztah replied:
If the CLI uses zero as equivalent of "not set", then probably `minimum` should be used (zero is  a valid value, but indicates "no limit").

ethan-haynes replied:
That makes sense. I will Git started on those changes :octocat:

thaJeztah commented on cli/compose/types/types.go:
```diff
@@ -296,6 +296,11 @@ type ServiceVolumeVolume struct {
 
+// ServiceVolumeTmpfs are options for a service volume of type tmpfs
+type ServiceVolumeTmpfs struct {
+	Size int64
```
This should probably be an `uint64`? Or is a signed integer used elsewhere?

ethan-haynes replied:
Chose to use int64 over unit64 because the mount struct looked to expect TmpfsOptions to contain it. Source:
https://github.com/moby/moby/blob/6415f1dcf5c79da853c604733c56b43ae3d18ced/api/types/mount/mount.go#L108

thaJeztah replied:
Thanks! Wonder if that was for a reason ("special `-1` value, e.g.) or just overlooked during review of the initial implementation. 🤔

ethan-haynes commented:
Thanks for the helpful feedback. I will make the necessary changes and update the PR.

ethan-haynes commented:
I believe that I addressed most of the concerns that @thaJeztah brought up in my latest PR, with the exception being the issue regarding the directory of the compose file being set when a source is not passed to a volume of type bind (using the long compose syntax). I added explicit checks for an empty source in the convert package, but by the time it is received there it is already set. I am happy to alter/fix the behavior in this PR if it is decided that the current functionality is not desired. Or we/I can make a separate issue for it and address it in a different PR. Thanks! 😄

dnephin commented:
Not sure why jenkins didn't run the integration suite.

dnephin commented on cli/compose/convert/volume.go:
```diff
@@ -85,6 +82,62 @@ func convertVolumeToMount(
+) (mount.Mount, error) {
+
+	switch volume.Type {
+	case "volume":
```
I think the old default (when type is not set) would be volume, so I guess this case should also have `""`

dnephin commented on cli/compose/convert/volume.go:
```diff
@@ -22,43 +22,40 @@ func Volumes(serviceVolumes []composetypes.ServiceVolumeConfig, stackVolumes vol
-	if volume.Type == "bind" && volume.Volume != nil {
-		return result, errors.New("volume options are incompatible with type bind")
+	if volume.Type != "tmpfs" {
+		result.Source = volume.Source
```
This isn't really necessary because `handleTmpfsToMount()` will error out in this case. I would remove this so we don't have to check it in two places.

thaJeztah commented:
ping @shin- for Docker Compose

thaJeztah commented:
Oh! before I forget; @ethan-haynes could you also take care of updating the compose-file documentation? Possibly needs to go in the `vnext-compose` branch; https://github.com/docker/docker.github.io/tree/vnext-compose/compose but @mistyhacks may be able to give guidance there 😊

mistyhacks commented:
Yes, that's correct if it applies to a not-yet-released feature of Compose.

ethan-haynes commented:
Sounds good. I'll go ahead and update the documentation with new examples for the 3.6v compose release. 📝
